disassemblers/ddisasm/ddisasmBBorRef.py

```diff
@@ -0,0 +1,305 @@
+import logging
+import optparse
+import subprocess
+import time
+from enum import Enum
+from pathlib import Path
+from typing import Optional
+
+import blocks_pb2
+import refInf_pb2 
+import gtirb
+from gtirb_capstone.instructions import GtirbInstructionDecoder
+
+logging.basicConfig(level = logging.INFO)
+
+
+class DdisasmError(Exception):
+    """
+    Exception raised when ddisasm fails
+    """
+    pass
+
+
+def disassemble(binary: Path, output: Path):
+    """
+    Run ddisasm on the given binary
+    
+    :param binary: the path to a binary to run ddisasm on
+    :param output: the path to the output
+    """
+    try:
+        p = subprocess.run(
+            ["ddisasm", binary, "--ir", output, "-j", "1"], check=True
+        )
+    except subprocess.CalledProcessError as e:
+        raise DdisasmError(f"ddisasm failed on {binary}: {e}")
+
+
+def addBB(block, pbFunc):
+    """
+    Add the given block to the given pb function
+    """
+    pbBB = pbFunc.bb.add()
+    pbBB.size = block.size
+    pbBB.va = block.address
+    pbBB.parent = pbFunc.va
+
+    logging.debug(
+        "The basic block address is 0x{0:x}, size is {1}".format(
+            pbBB.va, pbBB.size
+        )
+    )
+
+    decoder = GtirbInstructionDecoder(block.module.isa)
+
+    for instr in decoder.get_instructions(block):
+        instr.address
+        logging.debug("Instruction address is 0x{0:x}, size is {1}"\
+                .format(instr.address, instr.size))
+        pbInst = pbBB.instructions.add()
+        pbInst.va = instr.address
+        pbInst.size = instr.size
+
+    # get successors
+    for out_edge in block.outgoing_edges:
+        target_block = out_edge.target
+        if not isinstance(target_block, gtirb.ProxyBlock):
+            child = pbBB.child.add()
+            child.va = target_block.address
+            logging.debug("Successor of 0x{0:x}: 0x{1:x}".format(
+                block.address, target_block.address)
+            )
+
+
+def dumpBlocks(
+    ir: gtirb.IR,
+    output: str
+) -> None:
+    """
+    Create pb blocks for the given IR and dump the results to the given output
+    path as string.
+    """
+    for m in ir.modules:
+        logging.info("Module %s", m.name)
+
+        module = blocks_pb2.module()
+        for function, func_name in m.aux_data["functionNames"].data.items():
+            entries = m.aux_data.get("functionEntries").data[function]
+            logging.debug("function: %s", func_name)
+
+            # Pick the lowest address if there are multiple entries
+            entry_block = min(entries)
```

Comment:
It does not sort by address, and instead, it iterates the list once to determine the smallest element. Why do you ask? am I missing something..?

Reply:
Sorry, what I meant is that `entries` are code blocks if I remember correctly and I don't know offhand if the default ordering for code blocks is based on addresses. E.g. If it sorts based on UUID then this could be problematic.

Should this be `min(entries, key= lambda x: x.address)` ?